### PR TITLE
fix: use correct upload url for github

### PR DESCRIPTION
**Pull request by zckv**
Fix issue #658 

The tests are perfectible.

### CODE REVIEW

File: tests/unit/semantic_release/hvcs/test_github.py [hunk 1]
```diff
@@ -213,24 +213,16 @@ def test_pull_request_url(default_gh_client, pr_number):
...
-
 ############
 # Tests which need http response mocking
 ############
 
 
 github_matcher = re.compile(rf"^https://{Github.DEFAULT_DOMAIN}")
 github_api_matcher = re.compile(rf"^https://{Github.DEFAULT_API_DOMAIN}")
+github_upload_matcher = re.compile(
+    rf"^https://{Github.DEFAULT_API_DOMAIN.replace('api', 'uploads')}"
```
Comment (bernardcooke53) [inline tests/unit/semantic_release/hvcs/test_github.py:224]: I think this is cleaner if you add a constant to the `Github` class
```
DEFAULT_ASSET_UPLOAD_DOMAIN = DEFAULT_API_DOMAIN.replace("api", "uploads")
```
saves doing the replace everywhere

File: semantic_release/hvcs/github.py [hunk 2]
```diff
@@ -186,7 +186,10 @@ def asset_upload_url(self, release_id: str) -> str:
         :return: URL found to upload for a release
         """
         # https://docs.github.com/en/enterprise-server@3.5/rest/releases/assets#upload-a-release-asset ?
-        return f"{self.api_url}/repos/{self.owner}/{self.repo_name}/releases/{release_id}/assets"
+        response = self.session.get(
+            f"{self.api_url}/repos/{self.owner}/{self.repo_name}/releases/{release_id}",
+        )
```
Comment (bernardcooke53) [inline semantic_release/hvcs/github.py:191]: This introduces a new source of errors that didn't exist before - what if you get a 404?

File: tests/unit/semantic_release/hvcs/test_github.py [hunk 3]
```diff
@@ -539,16 +531,68 @@ def test_create_or_update_release_when_create_fails_and_no_release_for_tag(
...
+            f"releases/{release_id}/"
+            "assets{?name,label}"
+        ),
+        "status": "success",
+    }
+    with requests_mock.Mocker(session=default_gh_client.session) as m:
+        m.register_uri("GET", github_api_matcher, json=resp_payload, status_code=200)
+        assert default_gh_client.asset_upload_url(
+            release_id
+        ) == "https://uploads.{domain}/repos/{owner}/{repo}/releases/{release_id}/assets".format(
```
Comment (bernardcooke53) [inline tests/unit/semantic_release/hvcs/test_github.py:551]: Here could also be simpler if you could do `f"https://{domain}...".format(domain=Github.ASSET_UPLOAD_DOMAIN, ...)`

File: tests/unit/semantic_release/hvcs/test_github.py [hunk 4]
```diff
@@ -559,9 +603,10 @@ def test_upload_asset_succeeds(
             == True
         )
         assert m.called
-        assert len(m.request_history) == 1
-        assert m.last_request.method == "POST"
-        assert m.last_request.url == "{url}?{params}".format(
+        assert len(m.request_history) == 2
+        post_req = m.last_request.copy()  # Reference of m.last_request is not fix
```
Comment (bernardcooke53) [inline tests/unit/semantic_release/hvcs/test_github.py:607]: What issues were you seeing with `m.last_request`? Aren't we examining the content of the last request regardless? Just curious why we need to take a copy

**Review by bernardcooke53 — COMMENTED** [review]
I think the code is generally good, thank you for spotting the failed upload :pray:

Just wanted to ask a question around `last_request` in the tests, and also find out what issues you were facing with content-type assertions?

**Comment by bernardcooke53** [comment]
Not part of the changes you've made but you did highlight on the issue - perhaps the choice to exit 1 or not should be covered under [`--strict` mode](https://python-semantic-release.readthedocs.io/en/latest/strict_mode.html)?

**Comment by zckv** [comment]
> Not part of the changes you've made but you did highlight on the issue - perhaps the choice to exit 1 or not should be covered under [`--strict` mode](https://python-semantic-release.readthedocs.io/en/latest/strict_mode.html)?

As a user of the command "semantic release publish", I would like to have an exit code of 1 if the upload url is wrong and not any package is uploaded. So that the workflow is shown as failed. Maybe the --strict mode should differentiate between a 404 and a 422 on the POST request ?

File: tests/unit/semantic_release/hvcs/test_github.py [hunk 5]
```diff
@@ -559,12 +601,24 @@ def test_upload_asset_succeeds(
...
             url=default_gh_client.asset_upload_url(mock_release_id),
             params=urlencode(urlparams),
         )
+        # There are two requests made
+        # 1 - GET api.github.com -> get release info
+        # 2 - POST uploads.github.com -> Uploads assets
+        # call to m.last_request.url seems to change the m.last_request reference to the older request
+        # A bug of requests_mock ?
+        assert m.last_request.method == "GET"  # FIXME Response to PR 661
+        # ^ You shall not pass
```
Comment (zckv) [inline tests/unit/semantic_release/hvcs/test_github.py:621]: This can be deleted if the m.last_request "bug" is found and  corrected.

**Comment by bernardcooke53** [comment]
Hey @zckv - apologies for taking a while to come back to your comments, I've been sick for a couple of weeks and haven't really been able to look at this. I'll try and circle back shortly, re-review with your comments and finish this off

**Comment by zckv** [comment]
Hello @bernardcooke53,
I hope you are doing better. If further help is needed on this PR, I would like to try doing it if possible. And thank you for maintaining this helpful project.

**Comment by bernardcooke53** [comment]
Hey @zckv - just looked back over the PR and rebased it on some other changes, fixed linter errors with the new linter, etc. I found out what was happening with the `last_request` reference, let me try and summarise:

In your test code you had the following in [the test](https://github.com/zckv/python-semantic-release/blob/e4bfc74786273a410ebfbd828d5f42e465fa980f/tests/unit/semantic_release/hvcs/test_github.py#L589):

<details><summary>test_upload_asset_succeeds</summary>

```python3
        ...
        assert (
            default_gh_client.upload_asset(
                release_id=mock_release_id,
                file=example_changelog_md.resolve(),
                label=label,
            )
            is True
        )
        assert m.called
        assert len(m.request_history) == 2
        post_req = m.last_request  # Reference of m.last_request is not fix
        assert isinstance(m.last_request, requests_mock.request._RequestObjectProxy)
        assert m.last_request.method == "POST"  # FIXME Response to PR 661
        assert (
            m.last_request.method == "POST"
        )  # last request does not seem to be a generator
        assert m.last_request.url == "{url}?{params}".format(
            url=default_gh_client.asset_upload_url(mock_release_id),
            params=urlencode(urlparams),
        )
        # There are two requests made
        # 1 - GET api.github.com -> get release info
        # 2 - POST uploads.github.com -> Uploads assets
        # call to m.last_request.url seems to change the m.last_request reference to the older request
        # A bug of requests_mock ?
        assert m.last_request.method == "GET"  # FIXME Response to PR 661
        # ^ You shall not pass

        # Check if content-type header was correctly set according to
        # mimetypes - not retesting guessing functionality
        assert {
            "Content-Type": mimetypes.guess_type(
                example_changelog_md.resolve(), strict=False
            )[0]
            or "application/octet-stream"
        }.items() <= post_req.headers.items()
        assert post_req.body == example_changelog_md.read_bytes()
```

</details>

But I think specifically because of this assertion (line 603):

```python3
        assert m.last_request.url == "{url}?{params}".format(
            url=default_gh_client.asset_upload_url(mock_release_id),
            params=urlencode(urlparams),
        )
```

You were calling `default_gh_client.assert_upload_url` in order to make the assertion, and that was changing the `last_request` reference on the mocker because a new `GET` request was being make :slightly_smiling_face: 

I changed the test to use `request_history` which should get around this issue - [this example](https://requests-mock.readthedocs.io/en/latest/history.html#request-objects) showed how you can use the history like a list of requests, which I thought worked quite well. What do you think of the code now?

**Comment by zckv** [comment]
The revisions from [53e27d4](https://github.com/python-semantic-release/python-semantic-release/commit/53e27d42f362ead7f606c8649e610508a3633826) seem fine, it's more readable. 
 
 Thanks for explaining the issue with `default_gh_client.assert_upload_url` !
 
 I synced the fork with upstream commits, I don't know if that was necessary.

**Comment by bernardcooke53** [comment]
Thanks @zckv :tada:

**Pull request merged**